File: examples/readme_example/readmeexample.py
"""
Full version of the GitHub README.
"""
import asyncio
from pathlib import Path

from pyrogram import Client

from tgintegration import BotController
from tgintegration import Response

# This example uses the configuration of `config.ini` (see examples/README)
examples_dir = Path(__file__).parent.parent.absolute()
SESSION_NAME: str = "tgintegration_examples"


# This example uses the configuration of `config.ini` (see examples/README)
def create_client(session_name: str = SESSION_NAME) -> Client:
    client = Client(
        session_name=session_name,
        workdir=examples_dir,
        config_file=examples_dir / "config.ini",
    )
    client.load_config()
    return client


async def run_example(client: Client):
    controller = BotController(
        peer="@BotListBot",  # We are going to run tests on https://t.me/BotListBot
        client=client,
        max_wait=8,  # Maximum timeout for responses (optional)
        wait_consecutive=2,  # Minimum time to wait for more/consecutive messages (optional)
        raise_no_response=True,  # Raise `InvalidResponseError` when no response received (defaults to True)
        global_action_delay=2.5,  # Choosing a rather high delay so we can follow along in realtime (optional)
    )

    print("Clearing chat to start with a blank screen...")
    await controller.clear_chat()

    print("Sending /start and waiting for exactly 3 messages...")
    async with controller.collect(count=3) as response:  # type: Response
        await controller.send_command("start")

    assert response.num_messages == 3
    print("Three messages received, bundled as a `Response`.")
    assert response.messages[0].sticker
    print("First message is a sticker.")

    print("Let's examine the buttons in the response...")
    inline_keyboard = response.inline_keyboards[0]
    assert len(inline_keyboard.rows[0]) == 3
    print("Yep, there are three buttons in the first row.")

    # We can also press the inline keyboard buttons, in this case based on a pattern:
    print("Clicking the button matching the regex r'.*Examples'")
    examples = await inline_keyboard.click(pattern=r".*Examples")

    assert "Examples for contributing to the BotList" in examples.full_text
    # As the bot edits the message, `.click()` automatically listens for "message edited"
    # updates and returns the new state as `Response`.

    print("So what happens when we send an invalid query or the peer fails to respond?")
    from tgintegration import InvalidResponseError

    try:
        # The following instruction will raise an `InvalidResponseError` after
        # `controller.max_wait` seconds. This is because we passed `raise_no_response=True`
        # during controller initialization.
        print("Expecting unhandled command to raise InvalidResponseError...")
        async with controller.collect():
            await controller.send_command("ayylmao")
    except InvalidResponseError:
        print("Ok, raised as expected.")

    # If `raise_` is explicitly set to False, no exception is raised
    async with controller.collect(raise_=False) as response:  # type: Response
        print("Sending a message but expecting no reply...")
        await client.send_message(controller.peer_id, "Henlo Fren")

    # In this case, tgintegration will simply emit a warning, but you can still assert
    # that no response has been received by using the `is_empty` property.
    assert response.is_empty

    print("Success!")


if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(run_example(create_client()))

File: examples/readme/readme-example.ts
import { ChatController, InvalidResponseError } from "@tgintegration/core";
import { createClient } from "../utils.js";

/**
 * Full version of the GitHub README example.
 */
async function main() {
  const client = await createClient();

  // We are going to run tests on https://t.me/BotListBot (or any other bot)
  // Note: BotListBot might be down, replace with your target bot.
  const peer = "@BotListBot";

  const controller = new ChatController(client, peer, {
    globalActionDelay: 2500, // 2.5s delay to follow along
  });

  console.log("Clearing chat to start with a blank screen...");
  await controller.clearChat();

  console.log("Sending /start and waiting for exactly 3 messages...");
  const response = await controller.collect(
    {
      numMessages: 3,
      maxWait: 8000,
    },
    async () => {
      await controller.sendCommand("start");
    },
  );

  if (response.count !== 3) {
    throw new Error(`Expected 3 messages, got ${response.count}`);
  }
  console.log("Three messages received, bundled as a `Response`.");

  // First message is a sticker
  console.log(
    "First message is a sticker:",
    response.messages[0].media?.type === "sticker",
  );

  // Get the inline keyboard and examine its buttons
  const inlineKeyboard = response.inlineKeyboards[0];
  console.log(
    "Three buttons in the first row:",
    inlineKeyboard.buttons[0].length === 3,
  );

  // Click the Examples button
  console.log("Clicking Examples button...");
  const clickResult = await inlineKeyboard.click(/.*Examples/);
  console.log("Button click acknowledged:", clickResult.answered);

  console.log(
    "So what happens when we send an invalid query or the peer fails to respond?",
  );

  try {
    console.log("Expecting unhandled command to raise InvalidResponseError...");
    await controller.collect(
      {
        maxWait: 8000,
        throwOnTimeout: true,
      },
      async () => {
        await controller.sendCommand("ayylmao");
      },
    );
  } catch (e) {
    if (e instanceof InvalidResponseError) {
      console.log("Ok, raised as expected.");
    } else {
      throw e;
    }
  }

  // If `throwOnTimeout` is false (default), no exception is raised
  const noReplyResponse = await controller.collect(
    {
      maxWait: 3000,
      throwOnTimeout: false,
    },
    async () => {
      console.log("Sending a message but expecting no reply...");
      // Send directly via client if needed, or use controller
      await controller.client.sendText(controller.peer, "Henlo Fren");
    },
  );

  if (noReplyResponse.count === 0) {
    console.log("Success! No response received as expected.");
  }
}

await main();
